File: bottom-navigation/src/main/java/it/sephiroth/android/library/bottomnavigation/BottomNavigationFixedItemView.java
package it.sephiroth.android.library.bottomnavigation;

import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.support.v4.view.ViewCompat;
import android.view.animation.DecelerateInterpolator;
import android.view.animation.Interpolator;

import it.sephiroth.android.library.bottonnavigation.R;
import proguard.annotation.Keep;

import static android.util.Log.DEBUG;
import static android.util.Log.INFO;
import static it.sephiroth.android.library.bottomnavigation.MiscUtils.log;

/**
 * Created by alessandro on 4/3/16 at 10:55 PM.
 * Project: MaterialBottomNavigation
 */
@SuppressLint ("ViewConstructor")
public class BottomNavigationFixedItemView extends BottomNavigationItemViewAbstract {
    private static final String TAG = BottomNavigationFixedItemView.class.getSimpleName();
    private final int iconSize;
    private int centerY;
    private final Interpolator interpolator = new DecelerateInterpolator();
    private float textWidth;
    private long animationDuration;
    private final int colorActive;
    private final int colorInactive;

    private final int paddingTopActive;
    private final int paddingTopInactive;
    private final int paddingBottom;
    private final int paddingHorizontal;
    private final int textSizeInactive;

    private static final float TEXT_SCALE_ACTIVE = 1.1666666667f;
    private float canvasTextScale;
    private float iconTranslation;
    private int textCenterX;
    private int textCenterY;
    private int centerX;
    private float textX;
    private float textY;

    public BottomNavigationFixedItemView(final BottomNavigation parent, boolean expanded, final MenuParser.Menu menu) {
        super(parent, expanded, menu);

        final Resources res = getResources();
        this.paddingTopActive = res.getDimensionPixelSize(R.dimen.bbn_fixed_item_padding_top_active);
        this.paddingTopInactive = res.getDimensionPixelSize(R.dimen.bbn_fixed_item_padding_top_inactive);
        this.paddingBottom = res.getDimensionPixelSize(R.dimen.bbn_fixed_item_padding_bottom);
        this.paddingHorizontal = res.getDimensionPixelSize(R.dimen.bbn_fixed_item_padding_horizontal);
        this.textSizeInactive = res.getDimensionPixelSize(R.dimen.bbn_fixed_text_size_inactive);
        this.iconSize = res.getDimensionPixelSize(R.dimen.bbn_fixed_item_icon_size);

        this.animationDuration = menu.getItemAnimationDuration();
        this.colorActive = menu.getColorActive();
        this.colorInactive = menu.getColorInactive();
        this.centerY = paddingTopActive;
        this.canvasTextScale = expanded ? TEXT_SCALE_ACTIVE : 1f;
        this.iconTranslation = expanded ? 0 : (paddingTopInactive - paddingTopActive);

        log(TAG, DEBUG, "colors: %x, %x", colorInactive, colorActive);

        this.textPaint.setColor(Color.WHITE);
        this.textPaint.setHinting(Paint.HINTING_ON);
        this.textPaint.setLinearText(true);
        this.textPaint.setSubpixelText(true);
        this.textPaint.setTextSize(textSizeInactive);
        this.textPaint.setColor(expanded ? colorActive : colorInactive);
    }

    @Override
    protected void onStatusChanged(final boolean expanded, final int size, final boolean animate) {

        if (!animate) {
            updateLayoutOnAnimation(1, expanded);
            setIconTranslation(expanded ? 0 : (paddingTopInactive - paddingTopActive));
            return;
        }

        final AnimatorSet set = new AnimatorSet();
        set.setDuration(animationDuration);
        set.setInterpolator(interpolator);

        final ValueAnimator animator1 = ObjectAnimator.ofFloat(this, "textScale", expanded ? TEXT_SCALE_ACTIVE : 1);

        animator1.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(final ValueAnimator animation) {
                final float fraction = animation.getAnimatedFraction();
                updateLayoutOnAnimation(fraction, expanded);
            }
        });

        final ValueAnimator animator2 =
            ObjectAnimator.ofFloat(this, "iconTranslation", expanded ? 0 : (paddingTopInactive - paddingTopActive));

        set.playTogether(animator1, animator2);
        set.start();
    }

    private void updateLayoutOnAnimation(final float fraction, final boolean expanded) {
        if (expanded) {
            final int color = (Integer) evaluator.evaluate(fraction, colorInactive, colorActive);
            icon.setColorFilter(color, PorterDuff.Mode.SRC_ATOP);
            textPaint.setColor(color);
            icon.setAlpha(Color.alpha(color));
        } else {
            final int color = (Integer) evaluator.evaluate(fraction, colorActive, colorInactive);
            icon.setColorFilter(color, PorterDuff.Mode.SRC_ATOP);
            textPaint.setColor(color);
            icon.setAlpha(Color.alpha(color));
        }
    }

    @Override
    protected void onLayout(final boolean changed, final int left, final int top, final int right, final int bottom) {
        log(TAG, INFO, "onLayout(%b)", changed);

        super.onLayout(changed, left, top, right, bottom);

        if (null == this.icon) {
            this.icon = getItem().getIcon(getContext());
            this.icon.setColorFilter(isExpanded() ? colorActive : colorInactive, PorterDuff.Mode.SRC_ATOP);
            this.icon.setBounds(0, 0, iconSize, iconSize);
        }

        if (changed) {
            int w = right - left;
            centerX = (w - iconSize) / 2;
            icon.setBounds(centerX, centerY, centerX + iconSize, centerY + iconSize);

        }

        if (textDirty || changed) {
            measureText();
            textDirty = false;
        }
    }

    private void measureText() {
        log(TAG, INFO, "measureText");

        final int width = getWidth();
        final int height = getHeight();

        textWidth = textPaint.measureText(getItem().getTitle());
        textX = paddingHorizontal + (((width - paddingHorizontal * 2) - textWidth) / 2);
        textY = height - paddingBottom;
        textCenterX = width / 2;
        textCenterY = height - paddingBottom;
    }

    @Override
    protected void onDraw(final Canvas canvas) {
        super.onDraw(canvas);

        canvas.save();
        canvas.translate(0, iconTranslation);
        icon.draw(canvas);
        drawBadge(canvas);
        canvas.restore();

        canvas.save();
        canvas.scale(canvasTextScale, canvasTextScale, textCenterX, textCenterY);

        canvas.drawText(
            getItem().getTitle(),
            textX,
            textY,
            textPaint
        );

        canvas.restore();

    }

    @SuppressWarnings ("unused")
    @Keep
    public int getCenterY() {
        return centerY;
    }

    @SuppressWarnings ("unused")
    @Keep
    public void setCenterY(int value) {
        centerY = value;
        ViewCompat.postInvalidateOnAnimation(this);
    }

    @SuppressWarnings ("unused")
    @Keep
    public void setTextScale(final float value) {
        canvasTextScale = value;
        ViewCompat.postInvalidateOnAnimation(this);
    }

    @SuppressWarnings ("unused")
    @Keep
    public float getTextScale() {
        return canvasTextScale;
    }

    @Keep
    @SuppressWarnings ("unused")
    public void setIconTranslation(final float iconTranslation) {
        this.iconTranslation = iconTranslation;
        ViewCompat.postInvalidateOnAnimation(this);
    }

    @Keep
    @SuppressWarnings ("unused")
    public float getIconTranslation() {
        return iconTranslation;
    }

}

File: bottom-navigation/src/main/java/it/sephiroth/android/library/bottomnavigation/BottomNavigationFixedItemView.kt
package it.sephiroth.android.library.bottomnavigation

import android.annotation.SuppressLint
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PorterDuff
import android.view.animation.DecelerateInterpolator
import androidx.core.view.ViewCompat
import it.sephiroth.android.library.bottonnavigation.R
import timber.log.Timber

/**
 * Created by alessandro on 4/3/16 at 10:55 PM.
 * Project: MaterialBottomNavigation
 */
@Suppress("unused")
@SuppressLint("ViewConstructor")
internal class BottomNavigationFixedItemView(parent: BottomNavigation, expanded: Boolean, menu: MenuParser.Menu) :
        BottomNavigationItemViewAbstract(parent, expanded, menu) {
    private val iconSize: Int

    private var centerY: Int = 0
        set(value) {
            field = value
            ViewCompat.postInvalidateOnAnimation(this)
        }

    private val interpolator = DecelerateInterpolator()
    private var textWidth: Float = 0.toFloat()
    private var paddingBottomItem: Int
    private val paddingHorizontal: Int
    private val textSizeInactive: Int
    private var canvasTextScale: Float = 0.toFloat()

    private var iconTranslation: Float = 0.toFloat()
        set(value) {
            Timber.v("iconTranslation = $value")
            field = value
            ViewCompat.postInvalidateOnAnimation(this)
        }

    private var textCenterX: Int = 0
    private var textCenterY: Int = 0
    private var centerX: Int = 0
    private var textX: Float = 0.toFloat()
    private var textY: Float = 0.toFloat()

    var textScale: Float
        get() = canvasTextScale
        set(value) {
            Timber.v("textScale = $value")
            canvasTextScale = value
            ViewCompat.postInvalidateOnAnimation(this)
        }

    init {
        val res = resources
        this.paddingBottomItem = res.getDimensionPixelSize(R.dimen.bbn_fixed_item_padding_bottom)
        this.paddingHorizontal = res.getDimensionPixelSize(R.dimen.bbn_fixed_item_padding_horizontal)
        this.textSizeInactive = res.getDimensionPixelSize(R.dimen.bbn_default_text_size)
        this.iconSize = res.getDimensionPixelSize(R.dimen.bbn_default_item_icon_size)

        this.centerY = res.getDimensionPixelSize(R.dimen.bbn_fixed_item_padding_top)
        this.canvasTextScale = 1f
        this.iconTranslation = 0f
        this.textPaint.color = Color.WHITE
        this.textPaint.hinting = Paint.HINTING_ON
        this.textPaint.isLinearText = true
        this.textPaint.isSubpixelText = true
        this.textPaint.textSize = textSizeInactive.toFloat()
        this.textPaint.color = if (expanded) colorActive else colorInactive
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        this.textPaint.color =
                if (isExpanded) if (enabled) colorActive else colorDisabled else if (enabled) colorInactive else colorDisabled

        if (null != icon) {
            updateLayoutOnAnimation(1f, isExpanded)
        }
        requestLayout()
    }

    override fun onStatusChanged(expanded: Boolean, size: Int, animate: Boolean) {
        updateLayoutOnAnimation(1f, expanded)
    }

    private fun updateLayoutOnAnimation(fraction: Float, expanded: Boolean) {
        val enabled = isEnabled
        val dstColor = if (enabled) if (expanded) colorActive else colorInactive else colorDisabled
        val srcColor = if (enabled) if (expanded) colorInactive else colorActive else colorDisabled
        val color = evaluator.evaluate(fraction, srcColor, dstColor) as Int

        icon?.setColorFilter(color, PorterDuff.Mode.SRC_ATOP)
        icon?.alpha = Color.alpha(color)
        textPaint.color = color
        ViewCompat.postInvalidateOnAnimation(this)
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)

        if (null == this.icon) {
            this.icon = item!!.getIcon(context)!!.mutate()

            val color =
                    if (isExpanded) if (isEnabled) colorActive else colorDisabled else if (isEnabled) colorInactive else colorDisabled

            this.icon!!.setColorFilter(color, PorterDuff.Mode.SRC_ATOP)
            this.icon!!.setBounds(0, 0, iconSize, iconSize)
            this.icon!!.alpha = Color.alpha(color)
        }

        if (changed) {
            val w = right - left
            centerX = (w - iconSize) / 2
            icon?.setBounds(centerX, centerY, centerX + iconSize, centerY + iconSize)
        }

        if (textDirty || changed) {
            measureText()
            textDirty = false
        }
    }

    private fun measureText() {
        val width = width
        val height = height

        textWidth = textPaint.measureText(item!!.title)
        textX = paddingHorizontal + (width - paddingHorizontal * 2 - textWidth) / 2
        textY = (height - paddingBottomItem).toFloat()
        textCenterX = width / 2
        textCenterY = height - paddingBottomItem
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        canvas.save()
        canvas.translate(0f, iconTranslation)
        icon?.draw(canvas)
        drawBadge(canvas)
        canvas.restore()

        canvas.save()
        canvas.scale(canvasTextScale, canvasTextScale, textCenterX.toFloat(), textCenterY.toFloat())

        canvas.drawText(
                item!!.title,
                textX,
                textY,
                textPaint)

        canvas.restore()

    }

    companion object {
        private const val TEXT_SCALE_ACTIVE = 1.1666666667f
    }

}
